File: frontend/src/utils/recurrence.js
/**
 * Recurrence expansion (RFC-5545 subset) — anchored to the user's LOCAL
 * calendar, per the Calendar & Tasks plan.
 *
 * The backend stores recurring-task templates as an `rrule` string + UTC
 * `dtstart`; it does not expand them. Expansion happens here (and in the
 * mobile equivalent) using dayjs in the device timezone, so "every Mon/Wed/Fri"
 * means *local* Mon/Wed/Fri and occurrences land on the correct calendar day.
 *
 * Supported rule parts: FREQ (DAILY|WEEKLY|MONTHLY), INTERVAL, BYDAY (for
 * WEEKLY), UNTIL, COUNT. Anything else is ignored gracefully.
 */

import dayjs from "dayjs";

// RFC-5545 weekday tokens → dayjs day-of-week index (0 = Sunday).
const BYDAY_INDEX = { SU: 0, MO: 1, TU: 2, WE: 3, TH: 4, FR: 5, SA: 6 };

// Hard cap so an open-ended rule with a huge range can never loop forever.
const MAX_OCCURRENCES = 1000;

/**
 * Parse an rrule string ("FREQ=WEEKLY;BYDAY=MO,WE,FR;INTERVAL=1") into parts.
 * @param {string} rrule
 */
export const parseRrule = (rrule) => {
	const parts = {
		freq: null,
		interval: 1,
		byday: [],
		until: null,
		count: null,
	};
	if (!rrule || typeof rrule !== "string") return parts;
	for (const segment of rrule.split(";")) {
		const [rawKey, rawVal] = segment.split("=");
		if (!rawKey || !rawVal) continue;
		const key = rawKey.trim().toUpperCase();
		const val = rawVal.trim();
		switch (key) {
			case "FREQ":
				parts.freq = val.toUpperCase();
				break;
			case "INTERVAL": {
				const n = Number.parseInt(val, 10);
				if (Number.isFinite(n) && n > 0) parts.interval = n;
				break;
			}
			case "BYDAY":
				parts.byday = val
					.toUpperCase()
					.split(",")
					.map((d) => BYDAY_INDEX[d.trim()])
					.filter((i) => i !== undefined);
				break;
			case "UNTIL":
				parts.until = dayjs(val);
				break;
			case "COUNT": {
				const n = Number.parseInt(val, 10);
				if (Number.isFinite(n) && n > 0) parts.count = n;
				break;
			}
			default:
				break;
		}
	}
	return parts;
};

/**
 * Expand a recurrence into the local-date occurrences that fall within
 * [rangeStart, rangeEnd]. Each occurrence keeps the template's local time-of-day.
 *
 * @param {object} args
 * @param {string} args.rrule       RFC-5545 rule string
 * @param {string|Date} args.dtstart  UTC anchor (ISO string from backend)
 * @param {string|Date} args.rangeStart  visible window start
 * @param {string|Date} args.rangeEnd    visible window end
 * @param {string|Date|null} [args.until]  optional hard end (template.until)
 * @returns {Array<{ date: string, at: dayjs.Dayjs }>}  date = "YYYY-MM-DD" (local)
 */
export const expandRecurrence = ({
	rrule,
	dtstart,
	rangeStart,
	rangeEnd,
	until,
}) => {
	const rule = parseRrule(rrule);
	if (!rule.freq) return [];

	const start = dayjs(dtstart); // local representation of the UTC anchor
	if (!start.isValid()) return [];

	const windowStart = dayjs(rangeStart).startOf("day");
	const windowEnd = dayjs(rangeEnd).endOf("day");
	const hardUntil = until ? dayjs(until) : rule.until;

	// Iterate from `start`, never below the window, and stop at window/until/count.
	const out = [];
	let produced = 0;
	const hour = start.hour();
	const minute = start.minute();

	const withinHardLimits = (d) => {
		if (hardUntil?.isValid() && d.isAfter(hardUntil)) return false;
		if (rule.count != null && produced >= rule.count) return false;
		return true;
	};

	const push = (d) => {
		if (d.isBefore(windowStart) || d.isAfter(windowEnd)) return true; // keep walking
		out.push({ date: d.format("YYYY-MM-DD"), at: d });
		return true;
	};

	if (rule.freq === "WEEKLY") {
		const days = rule.byday.length ? rule.byday : [start.day()];
		// Walk week by week from the week containing `start`. Anchor on the Sunday
		// via `.day(0)` (pure Sunday-based arithmetic) rather than `startOf("week")`
		// so `weekAnchor.day(dow)` resolves each BYDAY token correctly regardless of
		// the global Monday week-start (see utils/dayjsConfig.js).
		let weekAnchor = start.day(0);
		let guard = 0;
		while (guard++ < MAX_OCCURRENCES * 2) {
			for (const dow of days) {
				const occ = weekAnchor.day(dow).hour(hour).minute(minute).second(0);
				if (occ.isBefore(start, "day")) continue; // before the template start
				if (!withinHardLimits(occ)) return out;
				produced += 1;
				push(occ);
				if (rule.count != null && produced >= rule.count) return out;
			}
			weekAnchor = weekAnchor.add(7 * rule.interval, "day");
			if (weekAnchor.isAfter(windowEnd) && weekAnchor.isAfter(start)) break;
			if (out.length >= MAX_OCCURRENCES) break;
		}
		return out;
	}

	// DAILY / MONTHLY: single step per interval from `start`.
	const stepUnit = rule.freq === "MONTHLY" ? "month" : "day";
	let cursor = start.hour(hour).minute(minute).second(0);
	let guard = 0;
	while (guard++ < MAX_OCCURRENCES) {
		if (!withinHardLimits(cursor)) break;
		if (cursor.isAfter(windowEnd)) break;
		produced += 1;
		push(cursor);
		cursor = cursor.add(rule.interval, stepUnit);
		if (out.length >= MAX_OCCURRENCES) break;
	}
	return out;
};

File: rust/src/apps/agenda/recurrence.rs
//! Server-side recurrence expansion (RFC-5545 subset), mirroring
//! `frontend/src/utils/recurrence.js` so the agenda endpoint can hand a
//! third-party agent the concrete recurring-task occurrences for a date range
//! in a single call — instead of returning bare rrule templates the caller
//! would have to expand itself.
//!
//! Expansion is done in **UTC** (dates only). The clients expand in the device
//! local timezone; for an API consumer the simplest precise contract is "pass a
//! UTC window, get UTC occurrence dates back". Supported parts: FREQ
//! (DAILY|WEEKLY|MONTHLY), INTERVAL, BYDAY (WEEKLY), UNTIL, COUNT.

use chrono::{DateTime, Datelike, Duration, Months, NaiveDate, NaiveDateTime, Utc};

// Hard cap so an open-ended rule with a huge range can never loop forever.
const MAX_OCCURRENCES: usize = 1000;

struct Rrule {
    freq: Option<String>,
    interval: u32,
    byday: Vec<u32>, // 0 = Sunday .. 6 = Saturday
    until: Option<NaiveDate>,
    count: Option<usize>,
}

// RFC-5545 weekday token → 0=Sun..6=Sat (matches chrono's num_days_from_sunday).
fn byday_index(tok: &str) -> Option<u32> {
    match tok.trim().to_uppercase().as_str() {
        "SU" => Some(0),
        "MO" => Some(1),
        "TU" => Some(2),
        "WE" => Some(3),
        "TH" => Some(4),
        "FR" => Some(5),
        "SA" => Some(6),
        _ => None,
    }
}

// UNTIL may be a date ("20261231"), a UTC datetime ("20261231T235959Z"), or an
// RFC-3339 instant. Reduce any of them to a UTC date.
fn parse_until(val: &str) -> Option<NaiveDate> {
    if let Ok(dt) = DateTime::parse_from_rfc3339(val) {
        return Some(dt.naive_utc().date());
    }
    for fmt in ["%Y%m%dT%H%M%SZ", "%Y%m%dT%H%M%S"] {
        if let Ok(dt) = NaiveDateTime::parse_from_str(val, fmt) {
            return Some(dt.date());
        }
    }
    for fmt in ["%Y-%m-%d", "%Y%m%d"] {
        if let Ok(d) = NaiveDate::parse_from_str(val, fmt) {
            return Some(d);
        }
    }
    None
}

fn parse_rrule(s: &str) -> Rrule {
    let mut r = Rrule {
        freq: None,
        interval: 1,
        byday: Vec::new(),
        until: None,
        count: None,
    };
    for segment in s.split(';') {
        let mut it = segment.splitn(2, '=');
        let key = it.next().unwrap_or("").trim().to_uppercase();
        let val = match it.next() {
            Some(v) => v.trim(),
            None => continue,
        };
        match key.as_str() {
            "FREQ" => r.freq = Some(val.to_uppercase()),
            "INTERVAL" => {
                if let Ok(n) = val.parse::<u32>() {
                    if n > 0 {
                        r.interval = n;
                    }
                }
            }
            "BYDAY" => r.byday = val.split(',').filter_map(byday_index).collect(),
            "UNTIL" => r.until = parse_until(val),
            "COUNT" => {
                if let Ok(n) = val.parse::<usize>() {
                    if n > 0 {
                        r.count = Some(n);
                    }
                }
            }
            _ => {}
        }
    }
    r
}

/// Expand a recurrence into the UTC dates that fall within
/// `[range_start, range_end]` (inclusive). `until` is the template's hard end,
/// which combines with any `UNTIL=` inside the rule (the earlier one wins).
pub fn expand_dates(
    rrule: &str,
    dtstart: DateTime<Utc>,
    until: Option<DateTime<Utc>>,
    range_start: NaiveDate,
    range_end: NaiveDate,
) -> Vec<NaiveDate> {
    let rule = parse_rrule(rrule);
    let Some(freq) = rule.freq.as_deref() else {
        return Vec::new();
    };

    let start = dtstart.date_naive();
    let hard_until = match (until.map(|u| u.date_naive()), rule.until) {
        (Some(a), Some(b)) => Some(a.min(b)),
        (a, b) => a.or(b),
    };
    let count = rule.count;

    let within_hard = |d: NaiveDate, produced: usize| -> bool {
        if let Some(hu) = hard_until {
            if d > hu {
                return false;
            }
        }
        if let Some(c) = count {
            if produced >= c {
                return false;
            }
        }
        true
    };

    let mut out: Vec<NaiveDate> = Vec::new();
    let mut produced: usize = 0;

    if freq == "WEEKLY" {
        let mut days = if rule.byday.is_empty() {
            vec![start.weekday().num_days_from_sunday()]
        } else {
            rule.byday.clone()
        };
        days.sort_unstable();
        days.dedup();

        // Week anchor = Sunday of the week containing `start` (dayjs default).
        let mut week_anchor = start - Duration::days(start.weekday().num_days_from_sunday() as i64);
        let mut guard = 0;
        while guard < MAX_OCCURRENCES * 2 {
            guard += 1;
            for &dow in &days {
                let occ = week_anchor + Duration::days(dow as i64);
                if occ < start {
                    continue; // before the template start
                }
                if !within_hard(occ, produced) {
                    return out;
                }
                produced += 1;
                if occ >= range_start && occ <= range_end {
                    out.push(occ);
                }
                if let Some(c) = count {
                    if produced >= c {
                        return out;
                    }
                }
            }
            week_anchor += Duration::days(7 * rule.interval as i64);
            if week_anchor > range_end && week_anchor > start {
                break;
            }
            if out.len() >= MAX_OCCURRENCES {
                break;
            }
        }
        return out;
    }

    // DAILY / MONTHLY: single step per interval from `start`.
    let mut cursor = start;
    let mut guard = 0;
    while guard < MAX_OCCURRENCES {
        guard += 1;
        if !within_hard(cursor, produced) {
            break;
        }
        if cursor > range_end {
            break;
        }
        produced += 1;
        if cursor >= range_start {
            out.push(cursor);
        }
        cursor = if freq == "MONTHLY" {
            match cursor.checked_add_months(Months::new(rule.interval)) {
                Some(d) => d,
                None => break,
            }
        } else {
            cursor + Duration::days(rule.interval as i64)
        };
        if out.len() >= MAX_OCCURRENCES {
            break;
        }
    }
    out
}
